src/OpenTelemetry/Metrics/DataPoint.cs

```diff
@@ -16,36 +16,85 @@
 
 using System;
 using System.Collections.Generic;
-
-#nullable enable
+using System.Runtime.CompilerServices;
 
 namespace OpenTelemetry.Metrics
 {
-    public abstract class DataPoint
+    internal struct DataPoint<T> : IDataPoint
```

Comment:
I think having `DataPoint<T>` be a `struct` here will be actually _worse_ for performance than having it be a `class`. Reason being...

```csharp
    public abstract class Aggregator
    {
        public virtual void Update(IDataPoint value)
    }
```

It is used everywhere as an `interface` so the struct is going to be boxed/copied to the heap. In that case, better to just create it on the heap directly.

Could get around that by doing stuff like this...

```csharp
    public abstract class Aggregator
    {
        public virtual void Update<T>(in T value)
            where T : struct, IDataPoint
    }
```

That will let the compiler treat it as an actual value type everywhere. But mutable structs are kind of evil, might be better off as a class anyway.

Reply:
Using existing code, if I just switch the struct to a class, the Benchmark shows it to be slower. It's also possible that I may of missed something.

As you pointed out, the issue is deciding when to take the T into a boxed version. We can chose to carry the T all the way to the aggregator (as you suggested). In current code, I took the boxing upfront (into a ThreadStatic to avoid allocation, and thus needing to have a "Reset" function instead of treating DataPoint as immutable).

My current goal is to avoid all alloc in the hot path. Presumably, this will also optimize on cycle performance.

Anyways, let's discuss and maybe we can come up with a cleaner/better solution.

Reply:
I'm marking this resolved. Please reopen if we want to move toward passing T fully to aggregators.

FYI: We have tried the approach of passing T fully down the pipe in our design-experiment branch some time ago. In that experiment, we found that approach was cumbersome (had to deal with T everywhere) and not necessarily faster either.

Reply:
Well, I guess I'll take this back. I tried passing T down the pipeline, and in this code base it was not bad at all. And performance is slightly better.